In [1]:
#Vamos a importar pandas
import pandas as pd

In [2]:
#Vamos a leer los datos
#Para lo anterior definimos una variable
#Cada vez que vamos a leer un archivo con PANDAS, siempre se deben posicionar nuestros datos en la misma carpeta que estamos trabajando, si no, tendremos que copiar la dirección completa
ca = pd.read_csv('water_potability.csv')
ca

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
0,NaN,204.890455,20791.318981,7.300212,368.516441,564.308654,10.379783,86.990970,2.963135,0
1,3.716080,129.422921,18630.057858,6.635246,NaN,592.885359,15.180013,56.329076,4.500656,0
2,8.099124,224.236259,19909.541732,9.275884,NaN,418.606213,16.868637,66.420093,3.055934,0
3,8.316766,214.373394,22018.417441,8.059332,356.886136,363.266516,18.436524,100.341674,4.628771,0
4,9.092223,181.101509,17978.986339,6.546600,310.135738,398.410813,11.558279,31.997993,4.075075,0
...,...,...,...,...,...,...,...,...,...,...
3271,4.668102,193.681735,47580.991603,7.166639,359.948574,526.424171,13.894419,66.687695,4.435821,1
3272,7.808856,193.553212,17329.802160,8.061362,NaN,392.449580,19.903225,NaN,2.798243,1
3273,9.419510,175.762646,33155.578218,7.350233,NaN,432.044783,11.039070,69.845400,3.298875,1
3274,5.126763,230.603758,11983.869376,6.303357,NaN,402.883113,11.168946,77.488213,4.708658,1


In [3]:
type(ca) #Así le preguntamos qué tipo  de datos es

pandas.core.frame.DataFrame

In [4]:
#VOY A CONFIGURAR PARA QUE ME MUESTRE TODAS LAS COLUMNAS
pd.set_option('display.max_columns', None)
ca

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
0,NaN,204.890455,20791.318981,7.300212,368.516441,564.308654,10.379783,86.990970,2.963135,0
1,3.716080,129.422921,18630.057858,6.635246,NaN,592.885359,15.180013,56.329076,4.500656,0
2,8.099124,224.236259,19909.541732,9.275884,NaN,418.606213,16.868637,66.420093,3.055934,0
3,8.316766,214.373394,22018.417441,8.059332,356.886136,363.266516,18.436524,100.341674,4.628771,0
4,9.092223,181.101509,17978.986339,6.546600,310.135738,398.410813,11.558279,31.997993,4.075075,0
...,...,...,...,...,...,...,...,...,...,...
3271,4.668102,193.681735,47580.991603,7.166639,359.948574,526.424171,13.894419,66.687695,4.435821,1
3272,7.808856,193.553212,17329.802160,8.061362,NaN,392.449580,19.903225,NaN,2.798243,1
3273,9.419510,175.762646,33155.578218,7.350233,NaN,432.044783,11.039070,69.845400,3.298875,1
3274,5.126763,230.603758,11983.869376,6.303357,NaN,402.883113,11.168946,77.488213,4.708658,1


In [6]:
#Para que me muestre las primeras cinco columnas
ca.head(5)

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
0,NaN,204.890455,20791.318981,7.300212,368.516441,564.308654,10.379783,86.990970,2.963135,0
1,3.716080,129.422921,18630.057858,6.635246,NaN,592.885359,15.180013,56.329076,4.500656,0
2,8.099124,224.236259,19909.541732,9.275884,NaN,418.606213,16.868637,66.420093,3.055934,0
3,8.316766,214.373394,22018.417441,8.059332,356.886136,363.266516,18.436524,100.341674,4.628771,0
4,9.092223,181.101509,17978.986339,6.546600,310.135738,398.410813,11.558279,31.997993,4.075075,0


In [7]:
#Para que me muestre las últimas cinco columnas
ca.tail()

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
3271,4.668102,193.681735,47580.991603,7.166639,359.948574,526.424171,13.894419,66.687695,4.435821,1
3272,7.808856,193.553212,17329.802160,8.061362,NaN,392.449580,19.903225,NaN,2.798243,1
3273,9.419510,175.762646,33155.578218,7.350233,NaN,432.044783,11.039070,69.845400,3.298875,1
3274,5.126763,230.603758,11983.869376,6.303357,NaN,402.883113,11.168946,77.488213,4.708658,1
3275,7.874671,195.102299,17404.177061,7.509306,NaN,327.459760,16.140368,78.698446,2.309149,1


In [8]:
#Mostramos la información de los datos
ca.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3276 entries, 0 to 3275
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ph               2785 non-null   float64
 1   Hardness         3276 non-null   float64
 2   Solids           3276 non-null   float64
 3   Chloramines      3276 non-null   float64
 4   Sulfate          2495 non-null   float64
 5   Conductivity     3276 non-null   float64
 6   Organic_carbon   3276 non-null   float64
 7   Trihalomethanes  3114 non-null   float64
 8   Turbidity        3276 non-null   float64
 9   Potability       3276 non-null   int64  
dtypes: float64(9), int64(1)
memory usage: 256.1 KB


In [10]:
#Verifico cuántos valores únicos tengo
ca.nunique()
#Observo que son valores cualitativos continuos
#Solamente la potabilidad tiene dos clases

ph                 2785
Hardness           3276
Solids             3276
Chloramines        3276
Sulfate            2495
Conductivity       3276
Organic_carbon     3276
Trihalomethanes    3114
Turbidity          3276
Potability            2
dtype: int64

In [9]:
#Imprimir los valores únicos o clases
for name in ca.columns.to_list():
    print(f'La columna:{name} contiene los elementos {ca[name ].unique()}')

La columna:ph contiene los elementos [       nan 3.71608008 8.09912419 ... 9.41951032 5.12676292 7.87467136]
La columna:Hardness contiene los elementos [204.89045547 129.42292051 224.23625939 ... 175.7626463  230.60375751
 195.10229859]
La columna:Solids contiene los elementos [20791.31898075 18630.05785797 19909.54173229 ... 33155.57821831
 11983.86937634 17404.17706105]
La columna:Chloramines contiene los elementos [7.30021187 6.63524588 9.2758836  ... 7.35023323 6.30335653 7.50930586]
La columna:Sulfate contiene los elementos [368.51644135          nan 356.88613564 ... 258.93060041 345.70025734
 359.94857437]
La columna:Conductivity contiene los elementos [564.30865417 592.88535913 418.60621306 ... 432.04478305 402.88311312
 327.45976046]
La columna:Organic_carbon contiene los elementos [10.37978308 15.18001312 16.86863693 ... 11.03906969 11.16894622
 16.14036763]
La columna:Trihalomethanes contiene los elementos [86.99097046 56.32907628 66.42009251 ... 69.84540029 77.4882131
 78.69

In [ ]:
#Dado que no hay variables con valores únicos, no se elimina ninguna columna

In [11]:
#Obtenemos el resumen estadístico
ca.describe()

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
count,2785.000000,3276.000000,3276.000000,3276.000000,2495.000000,3276.000000,3276.000000,3114.000000,3276.000000,3276.000000
mean,7.080795,196.369496,22014.092526,7.122277,333.775777,426.205111,14.284970,66.396293,3.966786,0.390110
std,1.594320,32.879761,8768.570828,1.583085,41.416840,80.824064,3.308162,16.175008,0.780382,0.487849
min,0.000000,47.432000,320.942611,0.352000,129.000000,181.483754,2.200000,0.738000,1.450000,0.000000
25%,6.093092,176.850538,15666.690297,6.127421,307.699498,365.734414,12.065801,55.844536,3.439711,0.000000
50%,7.036752,196.967627,20927.833607,7.130299,333.073546,421.884968,14.218338,66.622485,3.955028,0.000000
75%,8.062066,216.667456,27332.762127,8.114887,359.950170,481.792304,16.557652,77.337473,4.500320,1.000000
max,14.000000,323.124000,61227.196008,13.127000,481.030642,753.342620,28.300000,124.000000,6.739000,1.000000


In [12]:
#Obtengo la información para ls siguiente condición:
#Vamos a obtener cuántos valores son aguas duras y cuántos no
#Para ello ponemos la condición de que un agua dura es un ph mayor a 7.5, dureza > 180 mg/L y conductividad mayor a 300 µS/cm de acuerdo a la literatura
ca_aguas_duras = ca[(ca['ph' ] >7.5) & (ca['Hardness' ] >180) & (ca[ 'Conductivity']>350) ] 
ca_aguas_duras.describe()

#Se observa que hay 641 muestras que cumplen con esta condición, es decir 641 muestrras de las 3275 son de aguas duras o es lo mismo 19.57%


,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
count,641.000000,641.000000,641.000000,641.000000,492.000000,641.000000,641.000000,609.000000,641.000000,641.000000
mean,8.580929,214.849215,20898.891695,7.001712,330.631134,453.704505,14.411099,66.459684,3.919058,0.343214
std,0.926353,23.336968,8343.127618,1.561524,39.028046,67.480972,3.201112,16.642588,0.779319,0.475153
min,7.503086,180.060377,3413.081633,2.458609,129.000000,350.116686,4.371899,0.738000,1.844372,0.000000
25%,7.871961,198.354626,15049.584199,6.057868,309.843569,398.479317,12.397704,55.770690,3.381864,0.000000
50%,8.307463,211.663852,19909.541732,6.998962,332.294654,444.077017,14.525746,66.693973,3.892667,0.000000
75%,9.038326,226.284788,25694.873706,7.924616,356.204233,499.361690,16.575981,77.830816,4.477960,1.000000
max,14.000000,317.338124,56488.672413,13.127000,442.761428,753.342620,22.846529,124.000000,6.494749,1.000000


# Conclusiones
Se observa que se tiene una muestra de 3275 datos en total para diferentes valores de calidad química y biológica del agua, para este caso particular se ha buscado analizar la calidad del agua clasificándola en aguas duras. Esto, ya que es un problema que se presenta en diversas regiones del mundo, principalmente en aquellas donde la fuente de abstecimiento es por pozos.
De acuerdo con lo analizado, para toda la muestra se tiene un promedio de ph de 7.08 (ph neutro), una dureza promedio de 196 (agua dura), y una conductividad promedio de 426 (agua dura). 
Una vez que se aplicó el filtro para las condicones de agua dura (un ph mayor a 7.5, dureza > 180 mg/L y conductividad mayor a 300 µS/cm de acuerdo a la literatura) se obtuvo que 641 muestra de las 3275 analizadas cumplen con esta condición, teniendo un promedio de aguas duras del 19.57% sobre la muestra total; asimismo, el ph promedio de éstas es de 8.58, la dureza promedio de 214 y una conductividadpromedio de 453.
Es interesante observar que el promedio de la potabilidad ( a pesar de ser una variable dicotómica) toma valores de 0.39 (para la muestra completa) y de 0.34 para la muestra de aguas duras, es decir, la mayoria de las muestras tomadas no son potables y aptas para consumo humano.